contracts/DecentHats.sol

```diff
@@ -0,0 +1,436 @@
+                // Create election module and set as eligiblity
+                _createTermedHatAndAccountAndMintAndStreams(
+                    params.hatsProtocol,
+                    topHatAccount,
+                    params.hatsModuleFactory.createHatsModule( // Create election module and set as eligibility
+                        params.hatsElectionEligibilityImplementation,
+                        params.hatsProtocol.getNextId(adminHatId),
+                        abi.encode(topHatId, uint256(0)), // [BALLOT_BOX_ID, ADMIN_HAT_ID]
+                        abi.encode(params.hats[i].termedParam.termEndDateTs),
+                        uint256(SALT)
+                    ),
+                    adminHatId,
+                    params.hats[i]
+                );
+            } else {
+                _createHatAndAccountAndMintAndStreams(
+                    params.hatsProtocol,
+                    params.registry,
+                    topHatAccount,
+                    params.hatsAccountImplementation,
+                    adminHatId,
+                    params.hats[i]
+                );
+            }
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        params.hatsProtocol.transferHat(topHatId, address(this), msg.sender);
+    }
+
+    /**
+     * @notice Creates a new role hat and any streams on it.
+     *
+     * @notice This contract should be enabled a module on the Safe for which the role is to be created, and disabled after.
+     *
+     * @dev In order for the module to be able to create hats on behalf of the Safe, the Safe must first
+     * transfer its top hat to this contract. This function transfers the top hat back to the Safe after
+     * creating the role hat.
+     *
+     * @dev The function simply calls `_createHatAndAccountAndMintAndStreams` and then transfers the top hat back to the Safe.
+     *
+     * @dev Stream funds on Roles are targeted at the hat's smart account. In order to withdraw funds from the stream, the
+     * hat's smart account must be the one call to `withdraw-` on the Sablier contract, setting the recipient arg to its wearer.
+     *
+     * @dev Role hat creation, minting, smart account creation and stream creation are handled here in order
+     * to avoid a race condition where not more than one active proposal to create a new role can exist at a time.
+     * See: https://github.com/decentdao/decent-interface/issues/2402
+     */
+    function createRoleHat(CreateRoleHatParams calldata params) external {
+        _createHatAndAccountAndMintAndStreams(
+            params.hatsProtocol,
+            params.registry,
+            params.topHatAccount,
+            params.hatsAccountImplementation,
+            params.adminHatId,
+            params.hat
+        );
+
+        params.hatsProtocol.transferHat(
+            params.topHatId,
+            address(this),
+            msg.sender
+        );
+    }
```

Comment:
I think the last thing to do in this contract, is clean up this standalone function.

@DarksightKellar this one is on you.

So, the reason why we need to transfer the top back back to the safe, is because _this_ contract is the current wearer of the top hat by the time this function is being called. We did that (transaction was created in the frontend and part of the bundle of txs in the proposal that this call is in) because this contract can't create a new hat on a tree unless it's an admin of the hat it's trying to create. Makes sense.

BUT, instead of transferring the top hat to this contract, why not utilize the fact that we've added this contract as a module on the Safe! As of right now, adding this contract as a module on the Safe (in the context of minting a new hat via this call), _is completely unnecessary_, _unless we actually call `execTransactionFromModule` in this function call_.

So let's do that instead of transferring the top hat around.

Please proxy the call to create a new hat through the Safe via `execTransactionFromModule`, and remove the `transferHat` instruction. In the frontend, no need to transfer the top hat to this contract.